File: src/main/java/kr/chuyong/springspigot/commands/BukkitCommandHandler.java
package kr.chuyong.springspigot.commands;

import kr.chuyong.springspigot.annotation.CommandMapping;
import org.bukkit.Bukkit;
import org.bukkit.command.CommandMap;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.aop.support.AopUtils;
import org.springframework.util.ReflectionUtils;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;

public class BukkitCommandHandler {
    private static final Logger logger = LoggerFactory.getLogger(BukkitCommandHandler.class);
    private static final HashMap<String, BukkitCommandImpl> mainCMD = new HashMap<>();
    protected static HashMap<Class<?>, Object> globalInstanceMap = new HashMap<>();

    public static void registerCommands(Object cls) {
        try {
            Class<?> commandClazz = AopUtils.getTargetClass(cls);
            CommandMapping baseConfig = null;
            if (commandClazz.isAnnotationPresent(CommandMapping.class)) {
                baseConfig = commandClazz.getAnnotation(CommandMapping.class);
                registerParentCommand(baseConfig);
            }
            for (Method mt : ReflectionUtils.getAllDeclaredMethods(commandClazz)) {
                CommandMapping at = mt.getAnnotation(CommandMapping.class);
                if (at != null) {
                    if (baseConfig == null) {
                        //parent가 없을때
                        registerChildCommands(at.value(), new String[]{at.child()}, CommandConfig.fromAnnotation(at), mt, cls);
                    } else {
                        //class parent가 있을때
                        registerChildCommands(baseConfig.value(), new String[]{baseConfig.child(), at.value(), at.child()}, CommandConfig.fromAnnotation(at), mt, cls);
                    }

                }
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    private static BukkitCommandImpl registerParentCommand(CommandMapping ano) {
        if (ano.value().equals("") && ano.child().equals(""))
            throw new RuntimeException("Cannot Register non-named class commands");
        if (mainCMD.containsKey(ano.value())) return mainCMD.get(ano.value());
        try {
            final Field bukkitCommandMap = Bukkit.getServer().getClass().getDeclaredField("commandMap");
            bukkitCommandMap.setAccessible(true);
            CommandMap commandMap = (CommandMap) bukkitCommandMap.get(Bukkit.getServer());

            BukkitCommandImpl a = new BukkitCommandImpl(ano.value(), SuperCommandConfig.fromAnnotation(ano));
            commandMap.register(ano.value(), a);

            mainCMD.put(ano.value(), a);
            if (a.getAliases().size() < ano.aliases().length) {
                a.setAliases(Arrays.asList(ano.aliases()));
            }
            return a;
        } catch (Exception ex) {
            throw new RuntimeException();
        }
    }

    private static BukkitCommandImpl registerChildCommands(String parentKey, String[] childKey, CommandConfig ano, Method mtd, Object cl) {
        System.out.println("Register parent: " + parentKey + " child: " + Arrays.toString(childKey));
        try {
            final Field bukkitCommandMap = Bukkit.getServer().getClass().getDeclaredField("commandMap");
            bukkitCommandMap.setAccessible(true);
            CommandMap commandMap = (CommandMap) bukkitCommandMap.get(Bukkit.getServer());
            BukkitCommandImpl bukkitCommand = mainCMD.get(parentKey);
            if (bukkitCommand == null) {
                bukkitCommand = new BukkitCommandImpl(parentKey);
                commandMap.register(parentKey, bukkitCommand);
                mainCMD.put(parentKey, bukkitCommand);
            }

            SubCommandContainer container = bukkitCommand.addCommand(childKey, ano, mtd, cl);
            Bukkit.getConsoleSender().sendMessage("§f§l[§6SpringSpigot§f§l] §a/" + container.getFullKey() + " §f§lCommand Successfully Initialized");
            return bukkitCommand;

//            if (bukkitCommand.getAliases().size() < ano.aliases().length) {
//                bukkitCommand.setAliases(Arrays.asList(ano.aliases()));
//            }

        } catch (Exception e) {
            e.printStackTrace();
            throw new RuntimeException();
        }
    }
}

File: src/main/kotlin/chuyong/springspigot/command/BukkitCommandHandler.kt
package chuyong.springspigot.command

import chuyong.springspigot.command.annotation.CommandExceptionHandler
import chuyong.springspigot.command.annotation.CommandMapping
import chuyong.springspigot.command.data.CommandConfig
import chuyong.springspigot.command.data.InvokeWrapper
import chuyong.springspigot.command.data.SuperCommandConfig
import org.bukkit.Bukkit
import org.bukkit.command.CommandMap
import org.slf4j.LoggerFactory
import org.springframework.aop.support.AopUtils
import org.springframework.util.ReflectionUtils
import java.lang.reflect.Method
import java.util.*

object BukkitCommandHandler {
    private val logger = LoggerFactory.getLogger(BukkitCommandHandler::class.java)
    private val mainCMD = HashMap<String, BukkitCommandImpl>()
    var globalInstanceMap = HashMap<Class<*>, Any?>()
    var exceptionHandlers = HashMap<Class<out Throwable>, InvokeWrapper>()
    fun registerAdvices(obj: Any?) {
        try {
            val commandClazz = AopUtils.getTargetClass(obj!!)
            for (mt in ReflectionUtils.getAllDeclaredMethods(commandClazz)) {
                if (mt.isAnnotationPresent(CommandExceptionHandler::class.java)) {
                    val annotation = mt.getAnnotation(
                        CommandExceptionHandler::class.java
                    )
                    for (aClass in annotation.value) {
                        exceptionHandlers[aClass.java] = InvokeWrapper(mt, obj, null)
                    }
                }
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    fun registerCommands(cls: Any) {
        try {
            val commandClazz = AopUtils.getTargetClass(cls)
            var baseConfig: CommandMapping? = null
            if (commandClazz.isAnnotationPresent(CommandMapping::class.java)) {
                baseConfig = commandClazz.getAnnotation(CommandMapping::class.java)
                registerParentCommand(baseConfig)
            }
            for (mt in ReflectionUtils.getAllDeclaredMethods(commandClazz)) {
                val at = mt.getAnnotation(CommandMapping::class.java)
                if (at != null) {
                    if (baseConfig == null) {
                        //parent가 없을때
                        registerChildCommands(at.value, arrayOf(at.child), CommandConfig.fromAnnotation(at), mt, cls)
                    } else {
                        //class parent가 있을때
                        registerChildCommands(
                            baseConfig.value,
                            arrayOf(baseConfig.child, at.value, at.child),
                            CommandConfig.fromAnnotation(at),
                            mt,
                            cls
                        )
                    }
                }
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    private fun registerParentCommand(ano: CommandMapping): BukkitCommandImpl? {
        if (ano.value == "" && ano.child == "") throw RuntimeException("Cannot Register non-named class commands")
        return if (mainCMD.containsKey(ano.value)) mainCMD[ano.value] else try {
            val bukkitCommandMap = Bukkit.getServer().javaClass.getDeclaredField("commandMap")
            bukkitCommandMap.isAccessible = true
            val commandMap = bukkitCommandMap[Bukkit.getServer()] as CommandMap
            val a = BukkitCommandImpl(ano.value, SuperCommandConfig.fromAnnotation(ano))
            commandMap.register(ano.value, a)
            mainCMD[ano.value] = a
            if (a.aliases.size < ano.aliases.size) {
                a.aliases = Arrays.asList(*ano.aliases)
            }
            a
        } catch (ex: Exception) {
            throw RuntimeException()
        }
    }

    private fun registerChildCommands(
        parentKey: String,
        childKey: Array<String>,
        ano: CommandConfig,
        mtd: Method,
        cl: Any,
    ): BukkitCommandImpl {
        println("Register parent: " + parentKey + " child: " + Arrays.toString(childKey))
        return try {
            val bukkitCommandMap = Bukkit.getServer().javaClass.getDeclaredField("commandMap")
            bukkitCommandMap.isAccessible = true
            val commandMap = bukkitCommandMap[Bukkit.getServer()] as CommandMap
            var bukkitCommand = mainCMD[parentKey]
            if (bukkitCommand == null) {
                bukkitCommand = BukkitCommandImpl(parentKey)
                commandMap.register(parentKey, bukkitCommand)
                mainCMD[parentKey] = bukkitCommand
            }
            val container = bukkitCommand.addCommand(childKey, ano, mtd, cl)
            Bukkit.getConsoleSender()
                .sendMessage("§f§l[§6SpringSpigot§f§l] §a/" + container.fullKey + " §f§lCommand Successfully Initialized")
            bukkitCommand

//            if (bukkitCommand.getAliases().size() < ano.aliases().length) {
//                bukkitCommand.setAliases(Arrays.asList(ano.aliases()));
//            }
        } catch (e: Exception) {
            e.printStackTrace()
            throw RuntimeException()
        }
    }
}
